File: flutter/lib/pantallas/seccion5_informacion.dart
import 'package:flutter/material.dart';

import '../estado.dart';
import '../secciones.dart';
import '../widgets/comunes.dart';

const urlImagen = 'https://picsum.photos/id/1018/800/500';

const _modosEscalado = [
  ('Recortar', BoxFit.cover),
  ('Ajustar', BoxFit.contain),
  ('Estirar', BoxFit.fill),
  ('Original', BoxFit.none),
];

class Seccion5Informacion extends StatelessWidget {
  const Seccion5Informacion({super.key});

  @override
  Widget build(BuildContext context) {
    return const PantallaSeccion(
      seccion: Seccion.informacion,
      elementos: [
        ElementoCatalogo(
          nombre: 'Textos con distintos estilos',
          componente: 'Text + TextTheme + Text.rich(TextSpan)',
          descripcion: 'La tipografía establece jerarquía: títulos grandes, cuerpo legible y '
              'etiquetas pequeñas. El tamaño de estos ejemplos depende de la opción elegida en la Sección 3.',
          demo: _Textos(),
        ),
        ElementoCatalogo(
          nombre: 'Imagen local e imagen desde URL',
          componente: 'Image.asset / Image.network',
          descripcion: 'La primera imagen está incluida en la aplicación y la segunda se descarga '
              'de internet. El modo de escalado define cómo se adapta la imagen a su contenedor.',
          demo: _Imagenes(),
        ),
        ElementoCatalogo(
          nombre: 'Indicadores de progreso',
          componente: 'LinearProgressIndicator / CircularProgressIndicator',
          descripcion: 'Informan que una tarea está en curso. El modo determinado muestra el '
              'porcentaje avanzado; el indeterminado solo indica que se está trabajando.',
          demo: _Progreso(),
        ),
        ElementoCatalogo(
          nombre: 'Toast y Snackbar',
          componente: 'OverlayEntry (toast propio) / SnackBar',
          descripcion: 'El toast es un aviso breve que desaparece solo; Flutter no lo incluye y se '
              'construyó con un Overlay. El snackbar aparece abajo y puede incluir una acción.',
          demo: _ToastSnackbar(),
        ),
        ElementoCatalogo(
          nombre: 'Diálogo de confirmación',
          componente: 'showDialog + AlertDialog',
          descripcion: 'Interrumpe al usuario para confirmar una acción importante o irreversible. '
              'Presenta un título, una explicación y botones para aceptar o cancelar.',
          demo: _Dialogo(),
        ),
        ElementoCatalogo(
          nombre: 'Hoja inferior',
          componente: 'showModalBottomSheet',
          descripcion: 'Panel que se desliza desde la parte inferior con opciones adicionales '
              'relacionadas con el contenido. Se cierra deslizándola hacia abajo o tocando fuera.',
          demo: _HojaInferior(),
        ),
        ElementoCatalogo(
          nombre: 'Tarjeta, separador y distintivo',
          componente: 'Card.outlined / Divider / Badge',
          descripcion: 'La tarjeta agrupa información relacionada, el separador divide contenido '
              'visualmente y el distintivo (badge) muestra un contador sobre un ícono.',
          demo: _TarjetaBadge(),
        ),
      ],
    );
  }
}

class _Textos extends StatefulWidget {
  const _Textos();

  @override
  State<_Textos> createState() => _TextosState();
}

class _TextosState extends State<_Textos> {
  bool _negrita = false, _cursiva = false, _subrayado = false;

  @override
  Widget build(BuildContext context) {
    final estado = EstadoScope.of(context);
    final escala = estado.tamanoTexto.escala;
    final t = Theme.of(context).textTheme;
    final esquema = Theme.of(context).colorScheme;

    TextStyle escalar(TextStyle? estilo) => estilo!.copyWith(fontSize: estilo.fontSize! * escala);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('Titular grande', style: escalar(t.headlineMedium)),
        Text('Título de sección', style: escalar(t.titleLarge)),
        Text('Texto de cuerpo para párrafos largos.', style: escalar(t.bodyLarge)),
        Text('ETIQUETA PEQUEÑA', style: escalar(t.labelSmall)),
        const SizedBox(height: 4),
        Text.rich(
          TextSpan(
            style: escalar(t.bodyMedium),
            children: [
              const TextSpan(text: 'Texto con '),
              const TextSpan(text: 'negrita', style: TextStyle(fontWeight: FontWeight.bold)),
              const TextSpan(text: ', '),
              const TextSpan(text: 'cursiva', style: TextStyle(fontStyle: FontStyle.italic)),
              const TextSpan(text: ', '),
              TextSpan(text: 'color', style: TextStyle(color: esquema.primary)),
              const TextSpan(text: ', '),
              const TextSpan(text: 'tachado', style: TextStyle(decoration: TextDecoration.lineThrough)),
              const TextSpan(text: ' y '),
              const TextSpan(text: 'monoespaciado', style: TextStyle(fontFamily: 'monospace')),
              const TextSpan(text: '.'),
            ],
          ),
        ),
        const Divider(height: 24),
        Text('Aplica énfasis a la frase de ejemplo:', style: t.labelLarge),
        const SizedBox(height: 8),
        Wrap(
          spacing: 8,
          children: [
            FilterChip(label: const Text('Negrita'), selected: _negrita, onSelected: (v) => setState(() => _negrita = v)),
            FilterChip(label: const Text('Cursiva'), selected: _cursiva, onSelected: (v) => setState(() => _cursiva = v)),
            FilterChip(
                label: const Text('Subrayado'), selected: _subrayado, onSelected: (v) => setState(() => _subrayado = v)),
          ],
        ),
        const SizedBox(height: 8),
        Text(
          'El diseño es cómo funciona.',
          style: escalar(t.titleMedium).copyWith(
            fontWeight: _negrita ? FontWeight.bold : FontWeight.normal,
            fontStyle: _cursiva ? FontStyle.italic : FontStyle.normal,
            decoration: _subrayado ? TextDecoration.underline : TextDecoration.none,
          ),
        ),
        Row(
          children: [
            Expanded(child: TextoResultado('Tamaño actual: ${estado.tamanoTexto.etiqueta}')),
            TextButton(
              onPressed: () => irASeccion(context, Seccion.seleccion.ruta),
              child: const Text('Cambiar en Sección 3'),
            ),
          ],
        ),
      ],
    );
  }
}

class _Imagenes extends StatefulWidget {
  const _Imagenes();

  @override
  State<_Imagenes> createState() => _ImagenesState();
}

class _ImagenesState extends State<_Imagenes> {
  int _modo = 0;

  @override
  Widget build(BuildContext context) {
    final ajuste = _modosEscalado[_modo].$2;
    final etiqueta = Theme.of(context).textTheme.labelMedium;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        SizedBox(
          width: double.infinity,
          child: SegmentedButton<int>(
            showSelectedIcon: false,
            segments: [
              for (var i = 0; i < _modosEscalado.length; i++)
                ButtonSegment(value: i, label: Text(_modosEscalado[i].$1, style: etiqueta)),
            ],
            selected: {_modo},
            onSelectionChanged: (s) => setState(() => _modo = s.first),
          ),
        ),
        espacio,
        Row(
          children: [
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  _MarcoImagen(
                    child: Image.asset(
                      'assets/paisaje.png',
                      fit: ajuste,
                      semanticLabel: 'Paisaje de montañas incluido en la app',
                    ),
                  ),
                  Text('Local (asset)', style: etiqueta),
                ],
              ),
            ),
            const SizedBox(width: 8),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  _MarcoImagen(
                    child: Image.network(
                      urlImagen,
                      fit: ajuste,
                      semanticLabel: 'Fotografía descargada de internet',
                      loadingBuilder: (context, hijo, progreso) => progreso == null
                          ? hijo
                          : const Center(child: SizedBox(width: 32, height: 32, child: CircularProgressIndicator())),
                      errorBuilder: (context, error, pila) => const Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [Icon(Icons.broken_image), Text('Sin conexión')],
                      ),
                    ),
                  ),
                  Text('Desde URL (red)', style: etiqueta),
                ],
              ),
            ),
          ],
        ),
        espacio,
        TextoResultado('Modo de escalado: ${_modosEscalado[_modo].$1}'),
      ],
    );
  }
}

class _MarcoImagen extends StatelessWidget {
  const _MarcoImagen({required this.child});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return ClipRRect(
      borderRadius: BorderRadius.circular(12),
      child: Container(
        height: 120,
        width: double.infinity,
        color: Theme.of(context).colorScheme.surfaceContainerHighest,
        child: child,
      ),
    );
  }
}

class _Progreso extends StatefulWidget {
  const _Progreso();

  @override
  State<_Progreso> createState() => _ProgresoState();
}

class _ProgresoState extends State<_Progreso> with SingleTickerProviderStateMixin {
  late final AnimationController _animacion = AnimationController(vsync: this, duration: const Duration(seconds: 3))
    ..addListener(() => setState(() {}))
    ..addStatusListener((estado) {
      if (estado == AnimationStatus.completed && mounted) {
        mostrarSnackbar(context, 'Descarga simulada completada');
      }
    });
  bool _indeterminado = true;

  @override
  void dispose() {
    _animacion.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final valor = _animacion.value;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('Determinado: ${(valor * 100).toInt()} %', style: Theme.of(context).textTheme.labelLarge),
        espacio,
        Row(
          children: [
            Expanded(child: LinearProgressIndicator(value: valor)),
            const SizedBox(width: 16),
            CircularProgressIndicator(value: valor),
          ],
        ),
        espacio,
        Wrap(
          spacing: 8,
          children: [
            FilledButton(
              onPressed: _animacion.isAnimating ? null : () => _animacion.forward(from: 0),
              child: const Text('Iniciar'),
            ),
            OutlinedButton(onPressed: () => _animacion.reset(), child: const Text('Reiniciar')),
          ],
        ),
        const Divider(height: 24),
        SwitchListTile(
          contentPadding: EdgeInsets.zero,
          title: const Text('Mostrar indeterminados'),
          value: _indeterminado,
          onChanged: (v) => setState(() => _indeterminado = v),
        ),
        if (_indeterminado)
          const Row(
            children: [
              Expanded(child: LinearProgressIndicator()),
              SizedBox(width: 16),
              CircularProgressIndicator(),
            ],
          ),
      ],
    );
  }
}

class _ToastSnackbar extends StatefulWidget {
  const _ToastSnackbar();

  @override
  State<_ToastSnackbar> createState() => _ToastSnackbarState();
}

class _ToastSnackbarState extends State<_ToastSnackbar> {
  bool _archivado = false;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Wrap(
          spacing: 8,
          runSpacing: 8,
          children: [
            OutlinedButton(
              onPressed: () => mostrarToast(context, 'Esto es un toast'),
              child: const Text('Mostrar toast'),
            ),
            FilledButton(
              onPressed: () {
                setState(() => _archivado = true);
                mostrarSnackbar(context, 'Conversación archivada', accion: 'Deshacer', alAccion: () {
                  if (mounted) setState(() => _archivado = false);
                });
              },
              child: const Text('Mostrar snackbar'),
            ),
          ],
        ),
        espacio,
        TextoResultado(_archivado ? 'Estado: archivada (pulsa Deshacer en el snackbar)' : 'Estado: en la bandeja'),
      ],
    );
  }
}

class _Dialogo extends StatefulWidget {
  const _Dialogo();

  @override
  State<_Dialogo> createState() => _DialogoState();
}

class _DialogoState extends State<_Dialogo> {
  String _resultado = 'Aún no se ha abierto el diálogo.';

  Future<void> _abrir() async {
    final confirmado = await showDialog<bool>(
      context: context,
      builder: (context) => AlertDialog(
        icon: const Icon(Icons.delete),
        title: const Text('¿Eliminar archivo?'),
        content: const Text('El archivo “reporte.pdf” se eliminará de forma permanente. Esta acción no se puede deshacer.'),
        actions: [
          TextButton(onPressed: () => Navigator.pop(context, false), child: const Text('Cancelar')),
          TextButton(onPressed: () => Navigator.pop(context, true), child: const Text('Eliminar')),
        ],
      ),
    );
    if (!mounted) return;
    setState(() {
      _resultado = switch (confirmado) {
        true => 'Confirmaste: archivo eliminado.',
        false => 'Cancelaste la eliminación.',
        null => 'Diálogo cerrado sin elegir.',
      };
    });
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        FilledButton.tonalIcon(onPressed: _abrir, icon: const Icon(Icons.delete), label: const Text('Eliminar archivo')),
        espacio,
        TextoResultado(_resultado),
      ],
    );
  }
}

class _HojaInferior extends StatefulWidget {
  const _HojaInferior();

  @override
  State<_HojaInferior> createState() => _HojaInferiorState();
}

class _HojaInferiorState extends State<_HojaInferior> {
  String _resultado = 'Ninguna opción elegida.';

  Future<void> _abrir() async {
    final opcion = await showModalBottomSheet<String>(
      context: context,
      showDragHandle: true,
      builder: (context) => SafeArea(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 24, vertical: 8),
              child: Text('Compartir foto', style: Theme.of(context).textTheme.titleMedium),
            ),
            for (final (texto, icono) in const [
              ('Compartir', Icons.share),
              ('Copiar enlace', Icons.content_copy),
              ('Descargar', Icons.download),
            ])
              ListTile(
                leading: Icon(icono),
                title: Text(texto),
                onTap: () => Navigator.pop(context, texto),
              ),
            const SizedBox(height: 16),
          ],
        ),
      ),
    );
    if (opcion != null && mounted) setState(() => _resultado = 'Elegiste: $opcion');
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        FilledButton(onPressed: _abrir, child: const Text('Abrir hoja inferior')),
        espacio,
        TextoResultado(_resultado),
      ],
    );
  }
}

class _TarjetaBadge extends StatefulWidget {
  const _TarjetaBadge();

  @override
  State<_TarjetaBadge> createState() => _TarjetaBadgeState();
}

class _TarjetaBadgeState extends State<_TarjetaBadge> {
  int _notificaciones = 3;

  @override
  Widget build(BuildContext context) {
    final texto = Theme.of(context).textTheme;
    return Card.outlined(
      margin: EdgeInsets.zero,
      child: Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(16),
            child: IntrinsicHeight(
              child: Row(
                children: [
                  Badge(
                    isLabelVisible: _notificaciones > 0,
                    label: Text(_notificaciones > 99 ? '99+' : '$_notificaciones'),
                    child: const Icon(Icons.notifications, size: 32),
                  ),
                  const VerticalDivider(width: 32),
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text('Notificaciones', style: texto.titleMedium),
                        Text(
                          _notificaciones == 0 ? 'Estás al día' : 'Tienes $_notificaciones sin leer',
                          style: texto.bodySmall?.copyWith(color: Theme.of(context).colorScheme.onSurfaceVariant),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
            ),
          ),
          const Divider(height: 1),
          Padding(
            padding: const EdgeInsets.all(8),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.end,
              children: [
                TextButton(onPressed: () => setState(() => _notificaciones = 0), child: const Text('Marcar como leídas')),
                TextButton(onPressed: () => setState(() => _notificaciones++), child: const Text('Recibir una')),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

File: android-compose/app/src/main/java/com/javiergamez/catalogoui/compose/secciones/Seccion5Informacion.kt
@file:OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)

package com.javiergamez.catalogoui.compose.secciones

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.BrokenImage
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Download
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.Share
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Badge
import androidx.compose.material3.BadgedBox
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilledTonalButton
import androidx.compose.material3.FilterChip
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.ListItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedCard
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.VerticalDivider
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import coil3.compose.SubcomposeAsyncImage
import com.javiergamez.catalogoui.compose.R
import com.javiergamez.catalogoui.compose.componentes.ElementoCatalogo
import com.javiergamez.catalogoui.compose.componentes.PantallaSeccion
import com.javiergamez.catalogoui.compose.componentes.TextoResultado
import com.javiergamez.catalogoui.compose.componentes.rememberAvisos
import com.javiergamez.catalogoui.compose.datos.CatalogoViewModel
import com.javiergamez.catalogoui.compose.navegacion.Seccion
import kotlinx.coroutines.launch

const val URL_IMAGEN = "https://picsum.photos/id/1018/800/500"

private val MODOS_ESCALADO = listOf(
    "Recortar" to ContentScale.Crop,
    "Ajustar" to ContentScale.Fit,
    "Estirar" to ContentScale.FillBounds,
    "Original" to ContentScale.None,
)

@Composable
fun Seccion5Informacion(vm: CatalogoViewModel, onIrASeleccion: () -> Unit) {
    val avisos = rememberAvisos()

    PantallaSeccion(Seccion.INFORMACION) {
        item {
            ElementoCatalogo(
                nombre = "Textos con distintos estilos",
                componente = "Text + MaterialTheme.typography + AnnotatedString",
                descripcion = "La tipografía establece jerarquía: títulos grandes, cuerpo legible y " +
                    "etiquetas pequeñas. El tamaño de estos ejemplos depende de la opción elegida en la Sección 3.",
            ) {
                val escala = vm.tamanoTexto.escala
                var negrita by rememberSaveable { mutableStateOf(false) }
                var cursiva by rememberSaveable { mutableStateOf(false) }
                var subrayado by rememberSaveable { mutableStateOf(false) }

                @Composable
                fun Muestra(texto: String, estilo: TextStyle) {
                    Text(texto, style = estilo.copy(fontSize = estilo.fontSize * escala))
                }
                Muestra("Titular grande", MaterialTheme.typography.headlineMedium)
                Muestra("Título de sección", MaterialTheme.typography.titleLarge)
                Muestra("Texto de cuerpo para párrafos largos.", MaterialTheme.typography.bodyLarge)
                Muestra("ETIQUETA PEQUEÑA", MaterialTheme.typography.labelSmall)
                Text(
                    buildAnnotatedString {
                        append("Texto con ")
                        withStyle(SpanStyle(fontWeight = FontWeight.Bold)) { append("negrita") }
                        append(", ")
                        withStyle(SpanStyle(fontStyle = FontStyle.Italic)) { append("cursiva") }
                        append(", ")
                        withStyle(SpanStyle(color = MaterialTheme.colorScheme.primary)) { append("color") }
                        append(", ")
                        withStyle(SpanStyle(textDecoration = TextDecoration.LineThrough)) { append("tachado") }
                        append(" y ")
                        withStyle(SpanStyle(fontFamily = FontFamily.Monospace)) { append("monoespaciado") }
                        append(".")
                    },
                    style = MaterialTheme.typography.bodyMedium.let { it.copy(fontSize = it.fontSize * escala) },
                )
                HorizontalDivider()
                Text("Aplica énfasis a la frase de ejemplo:", style = MaterialTheme.typography.labelLarge)
                FlowRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    FilterChip(selected = negrita, onClick = { negrita = !negrita }, label = { Text("Negrita") })
                    FilterChip(selected = cursiva, onClick = { cursiva = !cursiva }, label = { Text("Cursiva") })
                    FilterChip(selected = subrayado, onClick = { subrayado = !subrayado }, label = { Text("Subrayado") })
                }
                Text(
                    "El diseño es cómo funciona.",
                    style = MaterialTheme.typography.titleMedium.let { it.copy(fontSize = it.fontSize * escala) },
                    fontWeight = if (negrita) FontWeight.Bold else FontWeight.Normal,
                    fontStyle = if (cursiva) FontStyle.Italic else FontStyle.Normal,
                    textDecoration = if (subrayado) TextDecoration.Underline else TextDecoration.None,
                )
                Row(verticalAlignment = Alignment.CenterVertically) {
                    TextoResultado("Tamaño actual: ${vm.tamanoTexto.etiqueta}", Modifier.weight(1f))
                    TextButton(onClick = onIrASeleccion) { Text("Cambiar en Sección 3") }
                }
            }
        }

        item {
            ElementoCatalogo(
                nombre = "Imagen local e imagen desde URL",
                componente = "Image(painterResource) / SubcomposeAsyncImage (Coil)",
                descripcion = "La primera imagen está incluida en la aplicación y la segunda se descarga " +
                    "de internet. El modo de escalado define cómo se adapta la imagen a su contenedor.",
            ) {
                var modo by rememberSaveable { mutableIntStateOf(0) }
                val escalado = MODOS_ESCALADO[modo].second
                SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
                    MODOS_ESCALADO.forEachIndexed { i, (etiqueta, _) ->
                        SegmentedButton(
                            selected = modo == i,
                            onClick = { modo = i },
                            shape = SegmentedButtonDefaults.itemShape(i, MODOS_ESCALADO.size),
                            icon = {},
                        ) { Text(etiqueta, style = MaterialTheme.typography.labelMedium) }
                    }
                }
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    Column(Modifier.weight(1f)) {
                        MarcoImagen {
                            Image(
                                painter = painterResource(R.drawable.paisaje),
                                contentDescription = "Paisaje de montañas incluido en la app",
                                contentScale = escalado,
                                modifier = Modifier.fillMaxSize(),
                            )
                        }
                        Text("Local (drawable)", style = MaterialTheme.typography.labelMedium)
                    }
                    Column(Modifier.weight(1f)) {
                        MarcoImagen {
                            SubcomposeAsyncImage(
                                model = URL_IMAGEN,
                                contentDescription = "Fotografía descargada de internet",
                                contentScale = escalado,
                                modifier = Modifier.fillMaxSize(),
                                loading = {
                                    Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                                        CircularProgressIndicator(Modifier.size(32.dp))
                                    }
                                },
                                error = {
                                    Column(
                                        Modifier.fillMaxSize(),
                                        horizontalAlignment = Alignment.CenterHorizontally,
                                        verticalArrangement = Arrangement.Center,
                                    ) {
                                        Icon(Icons.Filled.BrokenImage, contentDescription = null)
                                        Text("Sin conexión", style = MaterialTheme.typography.labelSmall)
                                    }
                                },
                            )
                        }
                        Text("Desde URL (Coil)", style = MaterialTheme.typography.labelMedium)
                    }
                }
                TextoResultado("Modo de escalado: ${MODOS_ESCALADO[modo].first}")
            }
        }

        item {
            ElementoCatalogo(
                nombre = "Indicadores de progreso",
                componente = "LinearProgressIndicator / CircularProgressIndicator",
                descripcion = "Informan que una tarea está en curso. El modo determinado muestra el " +
                    "porcentaje avanzado; el indeterminado solo indica que se está trabajando.",
            ) {
                val scope = rememberCoroutineScope()
                val progreso = remember { Animatable(0f) }
                var indeterminado by rememberSaveable { mutableStateOf(true) }
                Text("Determinado: ${(progreso.value * 100).toInt()} %", style = MaterialTheme.typography.labelLarge)
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(16.dp)) {
                    LinearProgressIndicator(progress = { progreso.value }, modifier = Modifier.weight(1f))
                    CircularProgressIndicator(progress = { progreso.value })
                }
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    Button(
                        onClick = {
                            scope.launch {
                                progreso.snapTo(0f)
                                progreso.animateTo(1f, tween(3000, easing = LinearEasing))
                                avisos.snackbar("Descarga simulada completada")
                            }
                        },
                        enabled = !progreso.isRunning,
                    ) { Text("Iniciar") }
                    OutlinedButton(onClick = { scope.launch { progreso.snapTo(0f) } }) { Text("Reiniciar") }
                }
                HorizontalDivider()
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text("Mostrar indeterminados", style = MaterialTheme.typography.labelLarge, modifier = Modifier.weight(1f))
                    Switch(checked = indeterminado, onCheckedChange = { indeterminado = it })
                }
                if (indeterminado) {
                    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(16.dp)) {
                        LinearProgressIndicator(modifier = Modifier.weight(1f))
                        CircularProgressIndicator()
                    }
                }
            }
        }

        item {
            ElementoCatalogo(
                nombre = "Toast y Snackbar",
                componente = "Toast (Android) / SnackbarHost",
                descripcion = "El toast es un aviso breve del sistema que desaparece solo. El snackbar " +
                    "aparece en la parte inferior de la app y puede incluir una acción, como deshacer.",
            ) {
                var archivado by rememberSaveable { mutableStateOf(false) }
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    OutlinedButton(onClick = { avisos.toast("Esto es un toast") }) { Text("Mostrar toast") }
                    Button(onClick = {
                        archivado = true
                        avisos.snackbar("Conversación archivada", "Deshacer") { archivado = false }
                    }) { Text("Mostrar snackbar") }
                }
                TextoResultado(if (archivado) "Estado: archivada (pulsa Deshacer en el snackbar)" else "Estado: en la bandeja")
            }
        }

        item {
            ElementoCatalogo(
                nombre = "Diálogo de confirmación",
                componente = "AlertDialog",
                descripcion = "Interrumpe al usuario para confirmar una acción importante o irreversible. " +
                    "Presenta un título, una explicación y botones para aceptar o cancelar.",
            ) {
                var mostrar by rememberSaveable { mutableStateOf(false) }
                var resultado by rememberSaveable { mutableStateOf("Aún no se ha abierto el diálogo.") }
                FilledTonalButton(onClick = { mostrar = true }) {
                    Icon(Icons.Filled.Delete, contentDescription = null, modifier = Modifier.size(18.dp))
                    Text("  Eliminar archivo")
                }
                TextoResultado(resultado)
                if (mostrar) {
                    AlertDialog(
                        onDismissRequest = {
                            mostrar = false
                            resultado = "Diálogo cerrado sin elegir."
                        },
                        icon = { Icon(Icons.Filled.Delete, contentDescription = null) },
                        title = { Text("¿Eliminar archivo?") },
                        text = { Text("El archivo “reporte.pdf” se eliminará de forma permanente. Esta acción no se puede deshacer.") },
                        confirmButton = {
                            TextButton(onClick = {
                                mostrar = false
                                resultado = "Confirmaste: archivo eliminado."
                            }) { Text("Eliminar") }
                        },
                        dismissButton = {
                            TextButton(onClick = {
                                mostrar = false
                                resultado = "Cancelaste la eliminación."
                            }) { Text("Cancelar") }
                        },
                    )
                }
            }
        }

        item {
            ElementoCatalogo(
                nombre = "Hoja inferior",
                componente = "ModalBottomSheet",
                descripcion = "Panel que se desliza desde la parte inferior con opciones adicionales " +
                    "relacionadas con el contenido. Se cierra deslizándola hacia abajo o tocando fuera.",
            ) {
                val estado = rememberModalBottomSheetState()
                val scope = rememberCoroutineScope()
                var mostrar by rememberSaveable { mutableStateOf(false) }
                var resultado by rememberSaveable { mutableStateOf("Ninguna opción elegida.") }
                Button(onClick = { mostrar = true }) { Text("Abrir hoja inferior") }
                TextoResultado(resultado)
                if (mostrar) {
                    ModalBottomSheet(onDismissRequest = { mostrar = false }, sheetState = estado) {
                        Text(
                            "Compartir foto",
                            style = MaterialTheme.typography.titleMedium,
                            modifier = Modifier.padding(horizontal = 24.dp, vertical = 8.dp),
                        )
                        listOf<Pair<String, ImageVector>>(
                            "Compartir" to Icons.Filled.Share,
                            "Copiar enlace" to Icons.Filled.ContentCopy,
                            "Descargar" to Icons.Filled.Download,
                        ).forEach { (opcion, icono) ->
                            ListItem(
                                headlineContent = { Text(opcion) },
                                leadingContent = { Icon(icono, contentDescription = null) },
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(horizontal = 8.dp)
                                    .clip(RoundedCornerShape(12.dp))
                                    .clickable {
                                        resultado = "Elegiste: $opcion"
                                        scope.launch { estado.hide() }.invokeOnCompletion { mostrar = false }
                                    },
                            )
                        }
                        Box(Modifier.height(24.dp))
                    }
                }
            }
        }

        item {
            ElementoCatalogo(
                nombre = "Tarjeta, separador y distintivo",
                componente = "OutlinedCard / HorizontalDivider / BadgedBox",
                descripcion = "La tarjeta agrupa información relacionada, el separador divide contenido " +
                    "visualmente y el distintivo (badge) muestra un contador sobre un ícono.",
            ) {
                var notificaciones by rememberSaveable { mutableIntStateOf(3) }
                OutlinedCard(Modifier.fillMaxWidth()) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier
                            .padding(16.dp)
                            .height(IntrinsicSize.Min),
                    ) {
                        BadgedBox(badge = {
                            if (notificaciones > 0) Badge { Text(if (notificaciones > 99) "99+" else "$notificaciones") }
                        }) {
                            Icon(Icons.Filled.Notifications, contentDescription = "Notificaciones", modifier = Modifier.size(32.dp))
                        }
                        VerticalDivider(Modifier.padding(horizontal = 16.dp).fillMaxHeight())
                        Column(Modifier.weight(1f)) {
                            Text("Notificaciones", style = MaterialTheme.typography.titleMedium)
                            Text(
                                if (notificaciones == 0) "Estás al día" else "Tienes $notificaciones sin leer",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                    }
                    HorizontalDivider()
                    Row(
                        horizontalArrangement = Arrangement.End,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(8.dp),
                    ) {
                        TextButton(onClick = { notificaciones = 0 }) { Text("Marcar como leídas") }
                        TextButton(onClick = { notificaciones++ }) { Text("Recibir una") }
                    }
                }
            }
        }
    }
}

@Composable
private fun MarcoImagen(contenido: @Composable () -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(120.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(MaterialTheme.colorScheme.surfaceContainerHighest),
    ) { contenido() }
}
